query.go

```diff
@@ -258,17 +248,33 @@ func (q *query) queryPeer(p peer.ID) *queryResult {
 
 		// add their addresses to the dialer's peerstore
 		q.dht.peerstore.AddAddrs(next.ID, next.Addrs, pstore.TempAddrTTL)
-		closer := q.localPeers.Add(next.ID)
-		foundCloserPeer = foundCloserPeer || closer
+		saw = append(saw, next.ID)
 	}
 
-	// did the successful query RPC fulfill the query stop condition ?
-	if q.stopFn(q.localPeers) {
-		q.cancel()
-	}
+	ch <- &queryUpdate{seen: saw, queried: []peer.ID{p}}
+}
 
-	return &queryResult{
-		foundCloserPeer: foundCloserPeer,
+func (q *query) populatePeers(up *queryUpdate) {
+	for _, p := range up.seen {
+		if p == q.dht.self { // don't add self.
+			continue
+		}
+		q.queryPeers.TryAdd(p)
+		q.queryPeers.MarkSeen(p)
+	}
+	for _, p := range up.queried {
+		if p == q.dht.self { // don't add self.
+			continue
+		}
+		q.queryPeers.TryAdd(p)
+		q.queryPeers.MarkQueried(p)
+	}
+	for _, p := range up.unreachable {
+		if p == q.dht.self { // don't add self.
+			continue
+		}
+		q.queryPeers.TryAdd(p)
+		q.queryPeers.MarkUnreachable(p)
```

Comment:
I've commented this elsewhere too:

We can simply remove the peer from `queryPeers` here  and not maintain this state at all.

Reply:
Two reasons:
* (Main) The QueryPeerset data structure does no support fast removes.
* Remembering the unreached peers allows me to add richer breakpoint conditions in code, while debugging.